naga/src/compact/functions.rs

```diff
@@ -30,10 +32,24 @@ impl FunctionTracer<'_> {
             }
         }
 
+        let mut typifier = crate::front::Typifier::new();
+
         // Treat named expressions as alive, for the sake of our test suite,
-        // which uses `let blah = expr;` to exercise lots of things.
+        // which uses `let blah = expr;` to exercise lots of things. The
+        // exception is expressions of abstract types, which can occur for local
+        // constants, which we must remove so that abstract types do not reach
+        // the IR.
         for (&value, _name) in &self.function.named_expressions {
-            self.expressions_used.insert(value);
+            typifier
+                .grow(value, &self.function.expressions, &self.resolve_context)
+                .expect("Failed to resolve type for named expression");
+            let mut expr_ty = typifier.get(value, self.types);
+            while let crate::TypeInner::Array { base, .. } = *expr_ty {
+                expr_ty = &self.types[base].inner;
+            }
```

Comment:
question: Here and in other places, we're recursing into types that can have abstract numeric components, and making sure those aren't abstract. This sounds correct. However, we're only checking for `Array`s' components. Don't we also need to check for vector and matrix components?

Reply:
Face-to-face discussion with @jamienicol: This is because the `scalar()` helper below handles the cases I'm thinking of, but not `Array`. A separate helper here sounds interesting.